Add menu support for native Windows

delta1186 opened the pull request:
References #146

I made a comment in the issue referenced above that gives a fairly detailed outline of changes made.  Please review that and my code and let me know what you would like for me to adjust.  I am sure there are probably better ways to code some of the array items.  Also, I extracted some code into private method where I thought it made sense.  Hopefully it readable.

I have manually tested all of the menus in native Windows 10.  I also manually tested the default menu system from the WSL prompt just to be sure my refactoring didn't break anything.  

I was reviewing the tests and noticed the default menu UI is not currently tested.  I think I could probably write a OS specific test to test the Windows menu.  Let me know if you want that to be included in this PR or just make a that a new PR.

Thanks!

delta1186 commented:
@mattstauffer and @josecanhelp I went ahead and rebased the PR to make sure it was still ready to merge.  If you have any questions or concerns, please don't hesitate to ask.  Thanks!

mattstauffer commented:
@delta1186 Thanks so much for this! Jose has the better Windows setup for testing, but he's slammed on a client project right now. We will get to this ASAP! 

Also, if anyone else in the community wants to pull this down and test it on their windows machine, I may be willing to just merge it without us testing. 

Thanks!

delta1186 commented:
@mattstauffer I went ahead and rebased my PR.  Of course it is currently broken.  Looks like some tests where added that are mocking the console which will not work on Windows. 

Also, I found an issue with the string format in Docker.php takeoutContainers method.  
https://github.com/tighten/takeout/blob/3f3d9656e701c2b5d581d7812036c22ef45fbf50/app/Shell/Docker.php#L91-L97
I am going to change it the following so that it runs on Windows.  I will verify mac and Linux don't break because of this.

```
$process = sprintf(
            'docker ps -a --filter "name=TO-" --format "table %s|%s"',
            '{{.ID}}|{{.Names}}|{{.Status}}|{{.Ports}}',
            '{{.Label \"com.tighten.takeout.Base_Alias\"}}|{{.Label \"com.tighten.takeout.Full_Alias\"}}'
        );
```
Want me to write a regression test for this?

In regards to the failing tests, how would you prefer I proceed?  
Option 1: Make menu test OS specific and just have something like $this->assertTrue(true) at the end or nothing and let the test show as Incomplete because it does not have any assertions.

Option 2: Update tests that currently assert against the menu to have a conditional that checks for Windows and asserts against the Windows menu system I created else use default assertions that currently exist.  

This should provide the tests coverage you would need in order tp merge this PR.  Just let me know your thoughts.

delta1186 commented:
@mattstauffer so I am not the best at testing, but I am giving it a try.  Would you recommend any other assertions in the isWindows section?  Trying not to overthink it.  Thanks!
```
function it_can_disable_a_service_from_menu()
    {
        $services = Collection::make([
            [
                'container_id' =>  $postgressId =  '1234',
                'names' => $postgressName = 'postgress',
            ],
            [
                'container_id' =>'12345',
                'names' => 'meilisearch',
            ],
        ]);

        $this->mock(Docker::class, function ($mock) use ($services, $postgressId) {
            $mock->shouldReceive('isInstalled')->andReturn(true);
            $mock->shouldReceive('isDockerServiceRunning')->andReturn(true);
            $mock->shouldReceive('takeoutContainers')->andReturn($services);
            $mock->shouldReceive('attachedVolumeName')
                ->with($postgressId)->andReturnNull()->once();
            $mock->shouldReceive('removeContainer')
                ->with($postgressId)->once();
        });

        if ($this->isWindows()) {
            $disableableServices = $services->mapWithKeys(function ($item) {
                return [$item['container_id'] => $item['names']];
            })->toArray();

            array_push($disableableServices, '<info>Exit</>');

            $this->artisan('disable')
                ->expectsChoice('Takeout containers to disable', $postgressName, $disableableServices)
                ->assertExitCode(0);
        } else {
            $menuMock = $this->mock(Menu::class, function ($mock) use ($postgressId) {
                $mock->shouldReceive('addLineBreak')->andReturnSelf();
                $mock->shouldReceive('setPadding')->andReturnSelf();
                $mock->shouldReceive('open')->andReturn($postgressId)->once();
            });

            Command::macro(
                'menu',
                function (string $title, array $options) use ($services, $menuMock) {
                    Assert::assertEquals('Services to disable', $title);
                    Assert::assertEquals(
                        $services->mapWithKeys(function ($container) {
                            return [$container['container_id'] => $container['names']];
                        })->toArray(),
                        $options
                    );
                    return $menuMock;
                }
            );

            $this->artisan('disable');
        }
    }
```

delta1186 commented:
Hooray! @mattstauffer and @josecanhelp I am back to green. :)  

Currently I have updated / added tests DisableCommandTest.php

- it_can_disable_a_service_from_menu
- it_will_try_to_stop_docker_service_if_no_containers_are_running

The stopDockerService test only tests that the disable command will show the correct menu.  It doesn't test the stopDockerService method which I don't think currently works, but doesn't throw any errors.

**Additional items**
- Updated the [BaseServiceTest::it_enables_services](https://github.com/tighten/takeout/pull/164/commits/a27653ca7eead1ff017b9ba1c0ec005a60d5fd68) to work with my PR.
- Refactored OS checking to use Environment class.  
- [Fixed string format bug in Docker::takeoutCountainers](https://github.com/tighten/takeout/pull/164/commits/b41064083efc7b67a34a1983b0407717e504df2e) **(I can extract this into an separate PR if you want, just let know)**

My goal is to add Windows menu test for StopCommand, StartCommand, and EnableCommand very soon so that we can get this PR merged....hopefully! :)

Thanks and let me know if you have any questions or concerns.

delta1186 commented:
Hey @mattstauffer me again! :)  I know that you requested @josecanhelp to review this PR, but I wanted you know know that I have added test coverage for all the new windows menu code and for the default menu.  I have also rebased this PR as of today so it would be an easy merge. ;)

But in all seriousness, please let me know if you need me to make any updates or you have any questions.  I appreciate the opportunity to contribute!  Thanks, Kevin

josecanhelp commented:
@delta1186 We REALLY appreciate the hard work on this. I'm so sorry it's taken me this long to jump in here. I'm going to start reviewing this today and I'll get back to you. Thank you!

delta1186 commented:
@josecanhelp no need to apologize. I appreciate the chance to contribute. Just let me know if you have any questions or concerns. 🤙🏻

josecanhelp commented:
@delta1186 I was able to pull this down and play with it. Seriously, **thank you** for the contribution. I love seeing Takeout working in the native command prompt.

I did notice that I was prompted if I wanted to stop the docker service when I brought down all of the takeout containers, but selecting "yes" had no effect. Was this just something you missed or did you have it working?

delta1186 commented:
@josecanhelp so to be honest I just refactored that block of code to mimic what was already there to make the menu work for windows.  

It is my understanding that the  $docker->stopDockerService() does not work yet.  I can look at PR #150 that addresses issue #112, but I might just create my own PR because I think there is more to solving the issue than what was submitted in that PR.  However, I could be wrong.

mattstauffer commented on app/Commands/DisableCommand.php:
```diff
@@ -67,29 +72,44 @@ public function disableByServiceName(string $service): void
                 return;
             case 1:
-                $serviceContainerId = $serviceMatches->flip()->first();
+                $this->disableByContainerId($serviceMatches->flip()->first());
```
Is there a reason to refactor this to move the actual disabling into the methods, rather than keeping it like it was before where each case set `$serviceContainerId` and then the bottom handled it?

delta1186 replied:
Not that I can tell.  Might have ended up like that as I was coming up with the PR.  I have updated the code and will push shortly.

mattstauffer commented on app/Commands/EnableCommand.php:
```diff
@@ -32,28 +38,90 @@ public function handle(Services $services): void
 
-        if (! $option) {
-            return;
+    private function windowsMenu($category = null): ?string
```
My gut says we'll want to refactor this out, but since you've tied it up nicely into a single private method, that's fine for now--thanks!

delta1186 replied:
I think I can probably do something with this in a future PR.

mattstauffer commented on app/Services/BaseService.php:
```diff
@@ -210,4 +210,13 @@ protected function containerName(): string
         return 'TO--' . $this->shortName() . '--' . $this->tag . $portTag;
     }
+
+    public function validateDockerRunTemplate($dockerRunTemplate): string
```
I think I'd probably name this something like `sanitize` instead of `validate`

delta1186 replied:
Done and will push changes shortly.

mattstauffer reviewed (changes requested):
Honestly, I can tell you've spent a *ton* of time on this and my eyes are glazing over a bit on the windows stuff. Let me just request a few small changes, then we'll merge and we can iterate on the Windows stuff if needed.

Either way, thank you *so much* for your hard work on this!

delta1186 commented:
@mattstauffer just pushed update and everything is still green.  Thanks for the feedback and hope this addresses your requests.

mattstauffer commented:
Thank you @delta1186!!!
